starlette/routing.py

```diff
@@ -121,12 +123,12 @@ def compile_path(
     idx = 0
     param_convertors = {}
     for match in PARAM_REGEX.finditer(path):
-        param_name, convertor_type = match.groups("str")
+        param_name, convertor_type, args = match.groups("str")
```

Comment:
If no parentheses are found, `args` is equal to empty string `""`.

If parentheses are found, `args` is equal to the whole parenthesized string (!).

To reproduce (it's just copy-pasted `compile_path` function with pprints inside)

```python
from pprint import pprint

PARAM_REGEX = re.compile(
    "{([a-zA-Z_][a-zA-Z0-9_]*)(:[a-zA-Z_][a-zA-Z0-9_]*)?((?:(?:\\(.*\\))|(?:\\[.*\\]))*)}"
)

def compile_path(path):
    path_regex = "^"
    path_format = ""
    duplicated_params = set()
    idx = 0
    param_convertors = {}
    for match in PARAM_REGEX.finditer(path):
        param_name, convertor_type, args = match.groups("str")
        convertor_type = convertor_type.lstrip(":")
        pprint([param_name, convertor_type, args])
        pprint([type(param_name), type(convertor_type), type(args)])
```

```
>>> compile_path("/{name:str}")
 ['name', 'str', '']
[<class 'str'>, <class 'str'>, <class 'str'>]

>>> compile_path("/{foo:bar([0-1]{2},[2-3]{3})}")
['foo', 'bar', '([0-1]{2},[2-3]{3})']
[<class 'str'>, <class 'str'>, <class 'str'>]
```

Therefore, you would need something like this:

```python
if args:
    # Strip the `()` and split by `,`
    normalized_args = args.strip("()").split(",")
    convertor = CONVERTOR_TYPES[convertor_type](*normalized_args)
else:
    convertor = CONVERTOR_TYPES[convertor_type]
```

---

This would also mean that `CONVERTOR_TYPES` is of type `dict[str, Union[Convertor, Type[Convertor]]]`, because I think the builtin convertors can already be instantiated singletons and also begs the question, whether should the newly added initializer of `Convertor` only accept one argument and that of type `list[str]`, but that is a minor detail compared to the above mentioned points.

cc @aminalaee

Reply:
In other words, the functionality is now broken, but since testing was done on a `RegexConvertor` with one initializer argument, it could not be detected so -  `"/regex/{param:regex([0-9]{3}-[0-9]{3}-[0-9]{4})}"` and the initialization looked like:

```
inst = RegexConvertor("([0-9]{3}-[0-9]{3}-[0-9]{4})")
```

And therefore, `inst.regex = "([0-9]{3}-[0-9]{3}-[0-9]{4})"` - notice the parentheses, which however still matched the test case!

```
>>> re.match("abc", "abc")
<re.Match object; span=(0, 3), match='abc'>
>>> re.match("(abc)", "abc")
<re.Match object; span=(0, 3), match='abc'>
```

Reply:
Thank you @bibajz I checked it and I confirmed the bug and fixed it.